File: analysis/xyz.py
class XYZ:
    def __init__(self):
        self.xyzs = []
        
    def load(self, filename):
        file = open(filename, 'r')
        line = file.readline()
        var = line
        while var is not None:
            xyz = XYZ_Single()
            num = int(var)
            xyz.load(file, num)
            self.xyzs.append(xyz)
            var = file.readline()
            if (len(var.split()) == 0): 
                break
        file.close()
        
    def save(self, filename):
        file = open(filename, 'w')
        for xyz in self.xyzs:
            xyz.write(file)
        file.close()

class XYZ_Single:
    def __init__(self):
        self.number = 0
        # For safari, this is time since start.
        self.comment = ''
        # This is an array of atomic symbols
        self.atoms = []
        # This is an array of the arrays of values for the atoms above.
        self.values = []
        
    def load(self, file, number):
        self.number = number
        # Reading in like this includes a \n at the end.
        self.comment = file.readline().split()[0]
        for i in range(number):
            line = file.readline().split()
            # this is the array of values for the atom on this line
            value = []
            for j in range(len(line) - 1):
                value.append(float(line[j + 1]))
            self.values.append(value)
            self.atoms.append(line[0])
    
    def write(self, file):
        file.write(str(self.number)+'\n')
        #Comment is read in with the formatting, so includes own /n
        file.write(self.comment)
        for i in range(self.number):
            line = self.atoms[i]
            value = self.values[i]
            for j in range(len(value)):
                line = line + ' ' + str(value[j])
            line = line + '\n'
            file.write(line)
        return

File: src/xyz.cpp
#include "xyz.h"
#include "string_utils.h"

void XYZ_Single::load(std::ifstream& input, int number)
{
    //Store the number
    this->number = number;
    //First line is comment
    getline(input, comment);

    values = new double*[number];
    atoms = new std::string[number];

    std::string line;

    //For number lines, we read the line.
    for(int i = 0; i<number; i++)
    {
        getline(input, line);
        std::vector<std::string> args = split(line);
        atoms[i] = args[0];
        max_row_index = args.size()-2;
        values[i] = to_double_array(args, 1, max_row_index);
    }
}

void XYZ_Single::save(std::ofstream& output)
{
    output << number << std::endl;
    output << comment << std::endl;
    for(int i = 0; i<number; i++)
    {
        output << atoms[i] << " ";
        int num = max_row_index;
        for(int j = 0; j<num; j++)
        {
            output << values[i][j];
            if(j < num-1) output << " ";
        }
        output << std::endl;
    }
}

void XYZ::load(std::ifstream& input)
{
    std::string line;
    while(getline(input, line))
    {
        int num = atoi(line.c_str());
        XYZ_Single xyz;
        xyz.load(input, num);
        xyzs.push_back(xyz);
    }
}

void XYZ::save(std::ofstream& output)
{
    for(XYZ_Single xyz: xyzs)
    {
        xyz.save(output);
    }
}